### DISCUSSION TITLE
Can you use the test data both in the training process and predicting process?

### DISCUSSION THREAD

**Discussion by Reetos** [mrdbourke/pytorch-deep-learning#364] [General]
In the jupyter notebook "01_pytorch_workflow.ipynb", the test data is used in training process and prediction was done on the same test data.

Are the predicted results the same in both cases?
And why are the predicted results stored in two different variables (i.e. test_pred and y_preds) ?

**Comment by PastorJordi** [comment]
Test data is being used as "validation" data during training. In other words, the model is trained (and its weights adjusted) using training data but its performance is assessed using unseen data by the model (test data in the notebook). 
If you do some kind of model selection / tuning / scheduling based on loss in validation data during training, it is safer to get a validation set, else by reusing training set as a validation set you may indirectly overfit your model. 
As an example, let's say I trained a cat/dog CNN classifier for 50 epochs, just splitting the data into train and test as it is done in the notebook. By chance, my test set gets slightly more golden retrievers than the average population (but this goes unnoticed because the actual label is either "cat" or "dog"). Then, after training, I get that the "best" epoch in terms of loss was the 44th, so I decide to retrieve the weights of that epoch in order to use "the best" version of model I had (it was actually the best in terms of loss when there are unusually high amount of golden retrievers). Then, if I evaluate my model in the same data I will have the impression that it works good / generalizes well. However, if I grab new data from wherever (with normal ratio of golden retrievers), the performance likely will be lower than expected by the latter evaluation.
I hope that mock up example helps building some intuition, else you can try reading other people with curated writing skills like [this](https://www.v7labs.com/blog/train-validation-test-set)
